experimental/sdk/webpubsub-socketio-extension/src/EIO/components/web-pubsub-connection-manager.ts

```diff
@@ -0,0 +1,204 @@
+import { WebPubSubExtensionOptions, debugModule } from "../../common/utils";
+import { ClientConnectionContext } from "./client-connection-context";
+import { WEBPUBSUB_CLIENT_CONNECTION_FILED_NAME, WEBPUBSUB_TRANSPORT_NAME } from "./constants";
+import { WebPubSubServiceClient } from "@azure/web-pubsub";
+import type { BaseServer } from "engine.io";
+import {
+  ConnectRequest as WebPubSubConnectRequest,
+  ConnectResponse as WebPubSubConnectResponse,
+  WebPubSubEventHandler,
+} from "@azure/web-pubsub-express";
+
+const debug = debugModule("wps-sio-ext:EIO:ConnectionManager");
+
+/**
+ * A `WebPubSubConnectionManager` instance is created for each Engine.IO server instance. It's designed to:
+ * 1. Manages all Azure Web PubSub client connections and keep them consistent with corresponding EIO clients.
+ * 2. Handle upstream invoke requests from AWPS and then translate them into Engine.IO behaviours.
+ * 3. Translates Engine.IO behaviours to AWPS behaviours like REST API calls.
+ * 4. Makes the EIO `sid` same as its corresponding Azure Web PubSub client connection id.
+ */
+export class WebPubSubConnectionManager {
+  /**
+   * Each `WebPubSubConnectionManager` instance is bound to a Engine.IO server instance and vice versa.
+   */
+  public linkedEioServer: BaseServer;
+
+  /**
+   * Client for connecting to a Web PubSub hub
+   */
+  public serviceClient: WebPubSubServiceClient;
+
+  /**
+   * Map from the `connectionId` of each client to its corresponding logical `ClientConnectionContext`.
+   */
+  private _clientConnections: Map<string, ClientConnectionContext> = new Map();
+
+  /**
+   * Handle upstream invoke requests from AWPS.
+   */
+  private _webPubSubEventHandler: WebPubSubEventHandler;
+
+  /**
+   * Options for Azure Web PubSub service.
+   */
+  private _webPubSubOptions: WebPubSubExtensionOptions;
+
+  /**
+   * In native Engine.IO, the `sid` of each EIO connnection is generated by server randomly.
+   * As for AWPS, it generates `ConnectionId` for each client.
+   * For each EIO connection, the extension enforces its `sid` in server side is same as the ConnectionId assigned by service.
+   * This array stores all `ConnectionId` which is generated by AWPS and is prepared to be assigned to EIO connection.
+   */
+  private _candidateSids: Array<string> = new Array();
+
+  constructor(server: BaseServer, options: WebPubSubExtensionOptions) {
+    debug("constructor");
+
+    if (!options.connectionString || options.connectionString === "") {
+      throw new Error("Valid connectionString is required");
+    }
+
+    if (!options.hub || options.hub === "") {
+      throw new Error("Valid hub is required");
+    }
+
+    this.linkedEioServer = server;
+    this._webPubSubOptions = options;
+    this.serviceClient = new WebPubSubServiceClient(
```

Comment:
serviceClient actually also support other credentials like AzureCredential, maybe we don't limit options to only allow connectionStrings later